Simplify config code (minor edits)

jdavcs opened the pull request:
Minor changes, details in commit messages.

(I will rebuild rst and sample yaml files when all these are merged, because, it is likely, some conflict resolution may need to happen first)

jdavcs committed:
Modify schema defaults and their loading

- Modify defaults in the schema so that correct values are loaded (add
  notes to description where applicable).
- When the default in schema was inconcsistent with the default loaded
  in `__init__.py`, use the latter.
- Replace `kwargs.get(foo, default)` with `self.foo` if default is set
  correctly in the schema.
- Remove redundant assignments and variables (there are still more
  left).
- The defalt for `citation_cache_lock_dir` should be `citations/locks`.
- In a few cases where defaults are NOT loaded in `__init__.py`, and the
  values are clearly meant to be suggested values, move those defaults
  into description.

jdavcs committed:
Modify schema defaults and their loading

- Modify defaults in the schema so that correct values are loaded (add
  notes to description where applicable).
- When the default in schema was inconcsistent with the default loaded
  in `__init__.py`, use the latter.
- Replace `kwargs.get(foo, default)` with `self.foo` if default is set
  correctly in the schema.
- Remove redundant assignments and variables (there are still more
  left).
- The defalt for `citation_cache_lock_dir` should be `citations/locks`.
- In a few cases where defaults are NOT loaded in `__init__.py`, and the
  values are clearly meant to be suggested values, move those defaults
  into description.

jdavcs committed:
Replace resolve_path() with os.path.join()

The function was redundant: it checked whether its argument was an
absolute path, and if not, called os.path.join. This is exactly what
os.path.join already does (python 3.x and 2.7)

The only benefit of having a separate function is the ability to add
additinal processing to such calls - so when we need that, we can add
it. Meanwhile, it removes unnecessary function calls + makes the code
slightly less complex.

The function is not called from outside this module (not to be confused
with the resolve_path method of the mixin class), but it's duplicated in
webapps/reports/config.py and webapps/tool_shed/config.py (which are not
addressed in this commit).

jdavcs committed:
Replace resolve_path() with os.path.join()

The function was redundant: it checked whether its argument was an
absolute path, and if not, called os.path.join. This is exactly what
os.path.join already does (python 3.x and 2.7)

The only benefit of having a separate function is the ability to add
additinal processing to such calls - so when we need that, we can add
it. Meanwhile, it removes unnecessary function calls + makes the code
slightly less complex.

The function is not called from outside this module (not to be confused
with the resolve_path method of the mixin class), but it's duplicated in
webapps/reports/config.py and webapps/tool_shed/config.py (which are not
addressed in this commit).

jdavcs committed:
Correct 2 errors in prev commit

jdavcs committed:
Simplify involucro, tool_dependency_dir config

Remove redundant code
Move schema default value into option description

jdavcs committed:
Shorten 2 very long lines in config_schema.yml

Also will help reformat this: https://docs.galaxyproject.org/en/master/admin/config.html#inactivity-box-content

jdavcs committed:
Simplify builds_file_path, len_file_path config

Remove redundant code
Move schema default value into option description

jdavcs commented:
@jmchilton please take a look at involucro/tool_dependency_dir assignment (I removed the check for 'none', but that could be a special case, maybe?)

jmchilton commented:
"tool_dependency_dir" can be "None" - especially from ini. It was the old default and still the easiest way to disable tool dependencies? 

- https://github.com/galaxyproject/galaxy/commit/9f7d921d5d387f0704db66f9e190688122be9170
- https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tool_util/deps/__init__.py#L51

Maybe we should handle this being a proper ``null`` value when coming from the YAML and we should document the option being null better in the config schema, but this does seem to be needed to me at first glance. If ``tool_dependency_dir`` is null, we may want to still use containers (in fact we probably do want to) so we can't use ``tool_dependency_dir``.

Maybe I'm wrong though.

jdavcs commented:
I see 2 main issues here:
1. Currently we do not have a way to set a config value to None (proper null) if there's a default specified. We should, and this example is a perfect use case: 'dependencies' probably should remain as the default (later resolved to <data_dir>/dependencies), but it should be possible to override it with null.
2. There could be older galaxies with 'none' (as a string) specified for that option (possibly, other options too). Our current setup will break these settings, which shouldn't happen

Here's a possible solution:

1. Allow user to override option defaults with none (case-insensitive, quotes or no quotes). We could add a string_as_none function to lib/util. We *could* specify a 'nullable' schema attribute (and use it along with default/type/desc...) and enable it by default for all options and explicitly disable it where it makes no sense (booleans and numeric types). Or we could say all is nullable, and expect that people won't do stupid things. Or we could enable it by default for all string types (which makes less sense, because we have 'any' type).

2. The above would take care of 'none' values in older configurations. However, if we have more potential cases like this one (not involving 'none'), we could convert such old values before processing them: just group stuff like `if foo=='old special value': foo='new way of specifying that special value` into a method called somewhere here https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L223
That way we'd "clean" all older configs *before* processing them (and the cleaning logic would be all grouped together).

What do you think? @jmchilton

jdavcs commented:
Rebased.

jdavcs committed:
Explicitly check if tool_dependency_dir is "none"

As per discussion here:
https://github.com/galaxyproject/galaxy/pull/8648#issuecomment-532200597

This should be replaced with a more generalized solution.

jdavcs commented on lib/galaxy/config/__init__.py:
```diff
@@ -432,6 +432,10 @@ def _process_config(self, kwargs):
         self.containers_resolvers_config_file = containers_resolvers_config_file
 
+        # tool_dependency_dir can be "none" (in old configs). If so, set it to schema default
+        if self.tool_dependency_dir and self.tool_dependency_dir.lower() == 'none':
```
@jmchilton This should take care or "none" values from old commits. I'll come up with a better/more generalized solution in a later PR (very soon). This, I hope, is a good enough solution for now. 
(`self.tool_dependency_dir` will be overwritten by whatever value is passed in kwargs, including a string "none", so this will catch it)

jdavcs replied:
Marking as resolved.

jdavcs committed:
Remove notice on conda_prefix limitation

As per discussion in code review.

jdavcs committed:
Increase # iterations in a potentially flaky test

This is just a hunch. The test shouldn't fail, and it doesn't fail
locally. It seems this might fix it (and point to the cause).

jdavcs committed:
Fix rebase error

jdavcs committed:
Fix rebase error

jdavcs committed:
Remove redundant call to kwargs

This is already assigned

natefoo commented on lib/galaxy/config/__init__.py:
```diff
@@ -435,20 +423,19 @@ def _process_config(self, kwargs):
-        if mulled_channels:
-            self.mulled_channels = [c.strip() for c in mulled_channels.split(',')]
+        # tool_dependency_dir can be "none" (in old configs). If so, set it to schema default
+        if self.tool_dependency_dir and self.tool_dependency_dir.lower() == 'none':
```
This could be `if string_as_bool_or_none(self.tool_dependency_dir) is None:` (`string_as_bool_or_none` would need to be imported from `galaxy.util` although I'm surprised we don't use it in here already).

nsoranzo replied:
@natefoo Seems a bit misleading to me to use `string_as_bool_or_none()` given that `self.tool_dependency_dir` is normally a path, not a boolean.

natefoo replied:
That's a good point.

natefoo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -817,7 +854,10 @@ mapping:
+        default: |
+          Please register only one account - we provide this service free of charge and have limited
+          computational resources. Multi-accounts are tracked and will be subjected to account
+          termination and data deletion.  
```
```suggestion
          termination and data deletion.
```

natefoo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -845,7 +885,10 @@ mapping:
+        default: |
+          Your account has not been activated yet.  Feel free to browse around and see what's
+          available, but you won't be able to upload data or run jobs until you have verified your
+          email address.  
```
```suggestion
          email address.
```

jdavcs committed:
Whitespace cleanup as per code review

Co-Authored-By: Nate Coraor <nate@bx.psu.edu>

jdavcs committed:
Whitespace cleanup as per code review

Co-Authored-By: Nate Coraor <nate@bx.psu.edu>
